### PR TITLE
Support for deferred proxies using JImplements

### LINKED ISSUE #653: JImplements requires JVM to be running
The JImplements interface is really nice - it gives overrides checking and automatic proxying for a much more pleasant interface when dealing with nested objects than the equivalent ``JProxy`` approach.

Unfortunately the ``jpype.JImplements`` currently requires the JVM to be running which can be awkward to guarantee at module definition time. I'd rather not nest my class definitions to allow me to guarantee that the JVM has been started, instead I'd like to defer the work that is done in ``JImplements`` until I try to construct an instance of my class.

I hacked together an approach using the existing public API tools to do precisely this:

```
import jpype as jp


def JImplementsDeferred(*jimplements_args, **jimplements_kwargs):
    def JProxyCreator(cls):
        # Store the __new__ method, we are going to override it for the first
        # construction of one of these cls instances. After that, we can use
        # the original class's __new__ method.
        cls.__unwrapped_new__ = cls.__new__

        def __fast_new__(cls, *args, **kwargs):
            self = cls.__unwrapped_new__(cls)
            self.__init__(*args, **kwargs)
            return self

        def __new__(cls, *args, **kwargs):
            # JImplements modifies the class's __init__ in-place.
            if getattr(cls, '__jimplements_called__', None) is None:
                jp.JImplements(*jimplements_args, **jimplements_kwargs)(cls)
                cls.__jimplements_called__ = True
            # Now put the old __new__ constructor back on this class (as
            # this doesn't need to be called again).
            cls.__new__ = cls.__fast_new__
            return cls.__new__(cls, *args, **kwargs)

        cls.__fast_new__ = __fast_new__
        cls.__new__ = __new__
        return cls
    return JProxyCreator
```

I can then use ``JImplementsDeferred`` in exactly the same way as ``jpype.JImplements`` (i.e. as a class decorator).

It basically hijacks the object constructor to do the ``jp.JImplements`` magic (which is mostly override checking and attaching a ``__javaproxy__`` attribute). After doing this once it then reverts back to using the originally defined object constructor, so there is no subsequent performance penalty. I was tempted to re-implement the ``__javaproxy__`` logic from the JImplements wrapper, but actually calling the JImplements decorator itself turned out to be fairly clean and is making use of the public JPype API.

Are there any obvious flaws to this approach? Is this something you'd like to see in jpype?

### PR DESCRIPTION
**Pull request by Thrameos**
In tonight's endless PR, we change the pattern for proxy creation so that classes can implement interfaces before the JVM is actually started.   Proxies created in this fashion are not actually checked until the first object using the Proxy is created, but are otherwise the same as normal proxies.

Fixes #653

### CODE REVIEW

File: jpype/_jproxy.py [hunk 1]
```diff
@@ -23,41 +23,47 @@
...
-
-    def new(tp, *args, **kwargs):
-        self = _jpype._JProxy.__new__(tp, None, actualIntf)
-        cls.__init__(self, *args, **kwargs)
-        return self
-    members['__new__'] = new
-
-    # Return the augmented class
-    return type("proxy.%s" % cls.__name__, (cls, _jpype._JProxy), members)
+    if _jpype.isStarted():
```
Comment (pelson) [inline jpype/_jproxy.py]: Given the conversation at #504 I'm not sure JVM is started is the right test for deciding whether we should be doing interface/override checking. I rather think explicit control is a good thing here.

File: jpype/_jproxy.py [hunk 2]
```diff
@@ -23,41 +23,47 @@
...
+            actualIntf = getattr(cls2, "_jinterfaces", None)
+            if actualIntf is not None:
+                return actualIntf
+            actualIntf = _convertInterfaces(intf)
+            _verifyOverrides(cls2, actualIntf)
+            setattr(cls2, "_jinterfaces", actualIntf)
+            return actualIntf
+
+        def new(tp, *args, **kwargs):
+            actualIntf = init(cls)
```
Comment (pelson) [inline jpype/_jproxy.py]: It would be good to only do this once, as we do with the non deferred case.

**Comment by pelson** [comment]
I also took a shot at this in #659.

File: jpype/_jproxy.py (hunk 1 above)
Comment (Thrameos) [reply jpype/_jproxy.py]: We do interface and override checking regardless.  If it is prior to starting then checking is done at the first create, else we execute it immediately.

File: jpype/_jproxy.py (hunk 2 above)
Comment (Thrameos) [reply jpype/_jproxy.py]: I believe line 49 ensures we do this only once.

**Pull request closed without merging**